File: engine/src/main/java/org/kigalisim/lang/interpret/NumberParseUtil.java
/**
 * Utility class for parsing numbers with flexible thousands and decimal separators.
 *
 * <p>This class provides functionality to parse number strings that may use either
 * comma (,) or period (.) as thousands separators or decimal separators, based on
 * context analysis of the input string. It handles European format (123.456,789)
 * and US format (123,456.789) intelligently.</p>
 *
 * @license BSD-3-Clause
 */

package org.kigalisim.lang.interpret;

import java.math.BigDecimal;

/**
 * Utility for parsing numbers with flexible thousands and decimal separators.
 */
public class NumberParseUtil {

  /**
   * Parse a number string with flexible thousands and decimal separators.
   *
   * <p>This method analyzes the input string to determine whether commas and periods
   * are being used as thousands separators or decimal separators. It supports both
   * European format (123.456,789) and US format (123,456.789).</p>
   *
   * <p>Logic:</p>
   * <ul>
   *   <li>Multiple occurrences of same character = thousands separator</li>
   *   <li>Single occurrence with ≠3 digits after = decimal separator</li>
   *   <li>Mixed separators: precedence rules apply (comma before period = comma thousands)</li>
   *   <li>Single occurrence with exactly 3 digits after = ambiguous (throws exception)</li>
   * </ul>
   *
   * @param numberString The number string to parse
   * @return BigDecimal representation of the parsed number
   * @throws NumberFormatException if the number format is ambiguous or invalid
   */
  public BigDecimal parseFlexibleNumber(String numberString) throws NumberFormatException {
    if (numberString == null || numberString.trim().isEmpty()) {
      throw new NumberFormatException("Number string cannot be null or empty");
    }

    // Clean whitespace and handle sign
    String cleanedInput = numberString.trim();
    boolean isNegative = cleanedInput.startsWith("-");
    boolean hasPositiveSign = cleanedInput.startsWith("+");

    if (isNegative || hasPositiveSign) {
      cleanedInput = cleanedInput.substring(1);
    }

    // Check for invalid start/end patterns
    // Allow numbers that start with decimal point like .5 but not with comma
    if (cleanedInput.startsWith(",") || cleanedInput.endsWith(",") || cleanedInput.endsWith(".")) {
      throw new NumberFormatException("Invalid number format: '" + numberString + "'");
    }

    // If no separators, parse directly
    if (!cleanedInput.contains(",") && !cleanedInput.contains(".")) {
      try {
        BigDecimal result = new BigDecimal(cleanedInput);
        return isNegative ? result.negate() : result;
      } catch (NumberFormatException e) {
        throw new NumberFormatException("Invalid number format: '" + numberString + "'");
      }
    }

    // Count occurrences of comma and period
    int commaCount = countOccurrences(cleanedInput, ',');
    int periodCount = countOccurrences(cleanedInput, '.');

    // Determine format and parse accordingly
    String parsedNumber = determineFormatAndParse(cleanedInput, commaCount, periodCount);

    try {
      BigDecimal result = new BigDecimal(parsedNumber);
      return isNegative ? result.negate() : result;
    } catch (NumberFormatException e) {
      throw new NumberFormatException("Invalid number format: '" + numberString + "'");
    }
  }

  /**
   * Determine the format of the number string and return a cleaned version for parsing.
   *
   * @param numberStr The number string (without sign)
   * @param commaCount Number of commas in the string
   * @param periodCount Number of periods in the string
   * @return Cleaned number string ready for BigDecimal parsing
   * @throws NumberFormatException if format is ambiguous or invalid
   */
  private String determineFormatAndParse(String numberStr, int commaCount, int periodCount)
      throws NumberFormatException {

    // Case 1: Only commas (no periods)
    if (commaCount > 0 && periodCount == 0) {
      return handleSingleSeparatorType(numberStr, ',', commaCount);
    }

    // Case 2: Only periods (no commas)
    if (periodCount > 0 && commaCount == 0) {
      return handleSingleSeparatorType(numberStr, '.', periodCount);
    }

    // Case 3: Both commas and periods present - apply precedence rules
    if (commaCount > 0 && periodCount > 0) {
      return handleMixedSeparators(numberStr);
    }

    // Should not reach here
    return numberStr;
  }

  /**
   * Handle numbers with only one type of separator (either comma or period).
   *
   * @param numberStr The number string
   * @param separator The separator character (',' or '.')
   * @param count Number of occurrences of the separator
   * @return Cleaned number string for parsing
   * @throws NumberFormatException if format is ambiguous
   */
  private String handleSingleSeparatorType(String numberStr, char separator, int count)
      throws NumberFormatException {

    if (count > 1) {
      // Multiple occurrences = thousands separator
      validateThousandsSeparatorPositions(numberStr, separator);
      return numberStr.replace(String.valueOf(separator), "");
    } else {
      // Single occurrence - check digit count after and position
      int separatorIndex = numberStr.indexOf(separator);
      int digitsAfter = numberStr.length() - separatorIndex - 1;
      int digitsBefore = separatorIndex;

      // Special case: if number starts with separator, it's always a decimal separator
      if (digitsBefore == 0) {
        if (separator == ',') {
          // Convert comma decimal to period decimal
          return numberStr.replace(",", ".");
        } else {
          // Already a period decimal
          return numberStr;
        }
      }

      if (digitsAfter == 3) {
        // Special case: if we have exactly 3 digits after AND the number before is >= 4 digits
        // OR if it's a common thousands separator pattern (like 1,000 or 10,000)
        // then treat it as thousands separator, not decimal
        if (digitsBefore >= 4 || isLikelyThousandsSeparator(numberStr, separatorIndex)) {
          return numberStr.replace(String.valueOf(separator), "");
        } else {
          // Truly ambiguous case - could be 123,456 (thousands) or 123.456 (decimal)
          throw new NumberFormatException(String.format(
              "Ambiguous number format: '%s'. Cannot determine if '%s' is a thousands "
              + "separator or decimal separator. Suggestions: Use '%s.0' for thousands "
              + "separator or change format to disambiguate.",
              numberStr, separator, numberStr
          ));
        }
      } else {
        // Not exactly 3 digits after = decimal separator
        if (separator == ',') {
          // Convert comma decimal to period decimal
          return numberStr.replace(",", ".");
        } else {
          // Already a period decimal
          return numberStr;
        }
      }
    }
  }

  /**
   * Determine if a single separator with 3 digits after is likely a thousands separator.
   *
   * @param numberStr The complete number string
   * @param separatorIndex The index of the separator
   * @return true if it's likely a thousands separator
   */
  private boolean isLikelyThousandsSeparator(String numberStr, int separatorIndex) {
    int digitsBefore = separatorIndex;

    // Common thousands patterns: 1,000  10,000  100,000 etc.
    // If digits before is 1-3 and we have exactly 3 zeros after, likely thousands
    if (digitsBefore >= 1 && digitsBefore <= 3) {
      String afterSeparator = numberStr.substring(separatorIndex + 1);
      // Check for common thousands patterns like 1,000 or 10,000
      if ("000".equals(afterSeparator)) {
        return true;
      }
    }

    // If we have 4 or more digits before the separator, it's likely thousands
    return digitsBefore >= 4;
  }

  /**
   * Handle numbers with both commas and periods present.
   *
   * @param numberStr The number string containing both separators
   * @return Cleaned number string for parsing
   * @throws NumberFormatException if format is invalid
   */
  private String handleMixedSeparators(String numberStr) throws NumberFormatException {
    int lastCommaIndex = numberStr.lastIndexOf(',');
    int lastPeriodIndex = numberStr.lastIndexOf('.');

    if (lastCommaIndex < lastPeriodIndex) {
      // Comma comes before period: comma = thousands, period = decimal (US format)
      // Example: 123,456.789
      String withoutThousands = numberStr.replace(",", "");
      validateSingleDecimalSeparator(withoutThousands, '.');
      // Also validate thousands separator positioning if any commas
      validateThousandsSeparatorPositions(numberStr.substring(0, lastPeriodIndex), ',');
      return withoutThousands;
    } else {
      // Period comes before comma: period = thousands, comma = decimal (European format)
      // Example: 123.456,789
      String withoutThousands = numberStr.replace(".", "");
      validateSingleDecimalSeparator(withoutThousands, ',');
      // Validate thousands separator positioning for periods
      validateThousandsSeparatorPositions(numberStr.substring(0, lastCommaIndex), '.');
      // Convert comma decimal to period decimal
      return withoutThousands.replace(",", ".");
    }
  }

  /**
   * Validate that thousands separators are positioned correctly.
   *
   * @param numberStr The number string
   * @param separator The thousands separator character
   * @throws NumberFormatException if positioning is invalid
   */
  private void validateThousandsSeparatorPositions(String numberStr, char separator)
      throws NumberFormatException {
    // Skip validation if no separators or empty string
    if (numberStr == null || numberStr.isEmpty() || !numberStr.contains(String.valueOf(separator))) {
      return;
    }

    String[] parts = numberStr.split("\\" + separator);

    // Must have at least 2 parts for a valid thousands separator
    if (parts.length < 2) {
      return;
    }

    // First part can be 1-3 digits
    if (parts[0].isEmpty() || parts[0].length() > 3) {
      throw new NumberFormatException("Invalid thousands separator format: '" + numberStr + "'");
    }

    // All subsequent parts must be exactly 3 digits
    for (int i = 1; i < parts.length; i++) {
      if (parts[i].length() != 3 || !parts[i].matches("\\d{3}")) {
        throw new NumberFormatException("Invalid thousands separator format: '" + numberStr + "'");
      }
    }
  }

  /**
   * Validate that there's only one decimal separator after removing thousands separators.
   *
   * @param numberStr The number string with thousands separators removed
   * @param decimalSeparator The expected decimal separator
   * @throws NumberFormatException if there are multiple decimal separators
   */
  private void validateSingleDecimalSeparator(String numberStr, char decimalSeparator)
      throws NumberFormatException {
    int count = countOccurrences(numberStr, decimalSeparator);
    if (count > 1) {
      throw new NumberFormatException("Invalid number format: multiple decimal separators in '"
          + numberStr + "'");
    }
  }

  /**
   * Count occurrences of a character in a string.
   *
   * @param str The string to search
   * @param ch The character to count
   * @return Number of occurrences
   */
  private int countOccurrences(String str, char ch) {
    int count = 0;
    for (int i = 0; i < str.length(); i++) {
      if (str.charAt(i) == ch) {
        count++;
      }
    }
    return count;
  }
}

File: editor/js/number_parse_util.js
/**
 * Utility class for parsing numbers with flexible thousands and decimal separators.
 * Mirrors the Java NumberParseUtil implementation to ensure consistency
 * between front-end and back-end.
 *
 * This class handles various number formats including:
 * - US format: "123,456.78"
 * - European format: "123.456,78"
 * - Mixed formats with proper precedence rules
 * - Detection of ambiguous formats like "123,456" vs "123.456"
 */

/**
 * Result class for number parsing operations that can succeed or fail.
 */
class NumberParseResult {
  /**
   * Create a NumberParseResult.
   *
   * @param {number|null} number - The parsed number, or null if parsing failed
   * @param {string|null} error - The error message, or null if parsing succeeded
   */
  constructor(number, error) {
    const self = this;
    self._number = number;
    self._error = error;
  }

  /**
   * Get the parsed number.
   *
   * @returns {number|null} The parsed number, or null if parsing failed
   */
  getNumber() {
    const self = this;
    return self._number;
  }

  /**
   * Get the error message.
   *
   * @returns {string|null} The error message, or null if parsing succeeded
   */
  getError() {
    const self = this;
    return self._error;
  }

  /**
   * Check if parsing was successful.
   *
   * @returns {boolean} True if parsing succeeded, false otherwise
   */
  isSuccess() {
    const self = this;
    return self._error === null;
  }

  /**
   * Create a successful result.
   *
   * @param {number} number - The parsed number
   * @returns {NumberParseResult} A successful result
   */
  static success(number) {
    return new NumberParseResult(number, null);
  }

  /**
   * Create a failed result.
   *
   * @param {string} error - The error message
   * @returns {NumberParseResult} A failed result
   */
  static error(error) {
    return new NumberParseResult(null, error);
  }
}
class NumberParseUtil {
  /**
   * Parse a number string with flexible thousands/decimal separator handling.
   *
   * @param {string} numberString - The number string to parse
   * @returns {NumberParseResult} The parsing result containing the number or error
   */
  parseFlexibleNumber(numberString) {
    const self = this;
    if (!numberString || typeof numberString !== "string") {
      return NumberParseResult.error("Number string cannot be null or empty");
    }

    const trimmed = numberString.trim();
    if (trimmed === "") {
      return NumberParseResult.error("Number string cannot be empty");
    }

    // Handle sign
    const isNegative = trimmed.startsWith("-");
    const isPositive = trimmed.startsWith("+");
    const numberPart = isNegative || isPositive ? trimmed.substring(1) : trimmed;

    // If no separators, parse directly
    const fullyNumeric = !numberPart.includes(",") && !numberPart.includes(".");
    if (fullyNumeric) {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${numberString}'`);
      }
      return NumberParseResult.success(isNegative ? -result : result);
    }

    // Handle cases with separators
    const separatorResult = self._parseWithSeparators(numberPart, numberString);
    if (!separatorResult.isSuccess()) {
      return separatorResult;
    }
    const finalNumber = isNegative ? -separatorResult.getNumber() : separatorResult.getNumber();
    return NumberParseResult.success(finalNumber);
  }

  /**
   * Check if a number string format is ambiguous.
   *
   * @param {string} numberString - The number string to check
   * @returns {boolean} True if the number format is ambiguous
   */
  isAmbiguous(numberString) {
    const self = this;
    if (!numberString || typeof numberString !== "string") {
      return false;
    }

    const result = self.parseFlexibleNumber(numberString);
    if (result.isSuccess()) {
      return false;
    }
    return result.getError().includes("Ambiguous number format");
  }

  /**
   * Get disambiguation suggestions for an ambiguous number.
   *
   * @param {string} numberString - The ambiguous number string
   * @returns {string} Suggestion message for resolving ambiguity
   */
  getDisambiguationSuggestion(numberString) {
    const self = this;
    if (!self.isAmbiguous(numberString)) {
      return "Number format is not ambiguous";
    }

    const trimmed = numberString.trim();
    const separator = trimmed.includes(",") ? "," : ".";
    const other = separator === "," ? "." : ",";

    return `Use '${trimmed}${separator}0' or '${trimmed}${other}0' to disambiguate`;
  }

  /**
   * Parse number string with comma and/or period separators.
   *
   * @param {string} numberPart - Number part without sign
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseWithSeparators(numberPart, originalString) {
    const self = this;
    const hasCommas = self._countOccurrences(numberPart, ",") > 0;
    const hasPeriods = self._countOccurrences(numberPart, ".") > 0;
    const bothSeparatorsPresent = hasCommas && hasPeriods;

    if (bothSeparatorsPresent) {
      return self._parseMixedSeparators(numberPart, originalString);
    } else if (hasCommas) {
      return self._parseSingleSeparatorType(numberPart, ",", originalString);
    } else if (hasPeriods) {
      return self._parseSingleSeparatorType(numberPart, ".", originalString);
    } else {
      // Shouldn't reach here based on calling conditions
      return NumberParseResult.error(`Unexpected parsing state for: '${originalString}'`);
    }
  }

  /**
   * Handle numbers with both comma and period separators.
   *
   * Validates that we have a valid mixed separator pattern:
   * - US format: One or more commas as thousands, one period as decimal
   *   (period must be last)
   * - European format: One or more periods as thousands, one comma as decimal
   *   (comma must be last)
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseMixedSeparators(numberPart, originalString) {
    const self = this;
    const commaCount = self._countOccurrences(numberPart, ",");
    const periodCount = self._countOccurrences(numberPart, ".");
    const lastComma = numberPart.lastIndexOf(",");
    const lastPeriod = numberPart.lastIndexOf(".");

    const usesUsFormat = lastPeriod > lastComma;
    const multiplePeriods = periodCount !== 1;
    const periodsBeforeComma = numberPart.indexOf(".") < lastComma;

    if (usesUsFormat) {
      if (multiplePeriods) {
        return NumberParseResult.error(
          `Invalid number format: '${originalString}' - multiple decimal separators not allowed`,
        );
      }

      if (periodsBeforeComma) {
        return NumberParseResult.error(
          `Invalid number format: '${originalString}' - periods cannot ` +
            "appear before commas in US format",
        );
      }

      const withoutThousands = numberPart.replace(/,/g, "");
      const result = parseFloat(withoutThousands);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    } else {
      const ukSuggestion = self._convertEuropeanToUkFormat(numberPart);
      return NumberParseResult.error(
        `Unsupported number format: '${originalString}'. ` +
        `Please use: '${ukSuggestion}'. ` +
        "Kigali Sim requires comma for thousands separator and period for decimal point.",
      );
    }
  }

  /**
   * Handle numbers with only one type of separator (all commas or all periods).
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseSingleSeparatorType(numberPart, separator, originalString) {
    const self = this;
    const separatorCount = self._countOccurrences(numberPart, separator);

    if (separatorCount > 1) {
      return self._parseSingleSeparatorMultipleOccurrences(
        numberPart,
        separator,
        originalString,
      );
    }

    const separatorIndex = numberPart.indexOf(separator);
    const digitsBefore = separatorIndex;
    const digitsAfter = numberPart.length - separatorIndex - 1;

    if (digitsAfter !== 3) {
      return self._parseSingleSeparatorSingleOccurrence(
        numberPart,
        separator,
        digitsBefore,
        digitsAfter,
        originalString,
      );
    }

    if (digitsBefore === 0) {
      return self._parseStartsWithSeparator(numberPart, separator, originalString);
    }

    if (digitsBefore >= 4) {
      return self._parseUnambiguousPriorDigits(numberPart, separator, originalString);
    }

    if (numberPart.startsWith("0" + separator)) {
      return self._parseWithLeadingZero(numberPart, separator, originalString);
    }

    return self._interpretAmbiguous(numberPart, separator, originalString);
  }

  /**
   * Parse single separator with multiple occurrences.
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseSingleSeparatorMultipleOccurrences(numberPart, separator, originalString) {
    const self = this;
    if (separator === ".") {
      const ukSuggestion = numberPart.replace(/\./g, ",");
      return NumberParseResult.error(
        self._generateEuropeanFormatError(originalString, ukSuggestion),
      );
    } else {
      const cleaned = numberPart.replace(new RegExp(`\\${separator}`, "g"), "");
      const result = parseFloat(cleaned);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Parse single separator with single occurrence and ambiguous digit count.
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {number} digitsBefore - Number of digits before separator
   * @param {number} digitsAfter - Number of digits after separator
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseSingleSeparatorSingleOccurrence(
    numberPart,
    separator,
    digitsBefore,
    digitsAfter,
    originalString,
  ) {
    const self = this;
    if (separator === ",") {
      if (digitsAfter <= 2 && digitsBefore > 0) {
        const ukSuggestion = numberPart.replace(",", ".");
        return NumberParseResult.error(
          self._generateEuropeanFormatError(originalString, ukSuggestion),
        );
      } else {
        const result = parseFloat(numberPart.replace(/,/g, ""));
        if (isNaN(result)) {
          return NumberParseResult.error(`Invalid number format: '${originalString}'`);
        }
        return NumberParseResult.success(result);
      }
    } else {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Parse number that starts with separator.
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseStartsWithSeparator(numberPart, separator, originalString) {
    const self = this;
    if (separator === ",") {
      const ukSuggestion = numberPart.replace(",", ".");
      return NumberParseResult.error(
        self._generateEuropeanFormatError(originalString, ukSuggestion),
      );
    } else {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Parse number with 4 or more digits before separator (unambiguous decimal).
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseUnambiguousPriorDigits(numberPart, separator, originalString) {
    const self = this;
    if (separator === ",") {
      const ukSuggestion = numberPart.replace(",", ".");
      return NumberParseResult.error(
        self._generateEuropeanFormatError(originalString, ukSuggestion),
      );
    } else {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Parse number starting with zero and separator (clearly a decimal).
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _parseWithLeadingZero(numberPart, separator, originalString) {
    const self = this;
    if (separator === ",") {
      const ukSuggestion = numberPart.replace(",", ".");
      return NumberParseResult.error(
        self._generateEuropeanFormatError(originalString, ukSuggestion),
      );
    } else {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Interpret ambiguous format with 3 digits after separator (like 123,456).
   *
   * @param {string} numberPart - Number part to parse
   * @param {string} separator - The separator character (',' or '.')
   * @param {string} originalString - Original string for error messages
   * @returns {NumberParseResult} Parsed number result
   * @private
   */
  _interpretAmbiguous(numberPart, separator, originalString) {
    const self = this;
    if (separator === ",") {
      const result = parseFloat(numberPart.replace(/,/g, ""));
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    } else {
      const result = parseFloat(numberPart);
      if (isNaN(result)) {
        return NumberParseResult.error(`Invalid number format: '${originalString}'`);
      }
      return NumberParseResult.success(result);
    }
  }

  /**
   * Check if a separator pattern looks like a thousands separator.
   *
   * @param {string} numberStr - The number string
   * @param {number} separatorIndex - Index of the separator
   * @returns {boolean} True if it looks like a thousands separator
   * @private
   */
  _isLikelyThousandsSeparator(numberStr, separatorIndex) {
    const self = this;
    const beforeSeparator = numberStr.substring(0, separatorIndex);

    // Use length-based check as suggested in feedback, but adjusted to handle common cases
    return beforeSeparator.length >= 1;
  }

  /**
   * Count occurrences of a character in a string.
   *
   * @param {string} str - The string to search
   * @param {string} char - The character to count
   * @returns {number} Number of occurrences
   * @private
   */
  _countOccurrences(str, char) {
    const self = this;
    return str.length - str.replaceAll(char, "").length;
  }

  /**
   * Convert European format number to UK format suggestion.
   *
   * Handle different European patterns:
   * - "1.234,56" to "1,234.56"
   * - "123,45" to "123.45"
   * - "1.234.567,89" to "1,234,567.89"
   *
   * @param {string} numberPart - European format number part
   * @returns {string} UK format equivalent
   * @private
   */
  _convertEuropeanToUkFormat(numberPart) {
    const lastCommaIndex = numberPart.lastIndexOf(",");
    const lastPeriodIndex = numberPart.lastIndexOf(".");

    const europeanMixed = lastCommaIndex > lastPeriodIndex && lastCommaIndex !== -1;
    const singleComma = lastCommaIndex !== -1 && lastPeriodIndex === -1;

    if (europeanMixed) {
      return numberPart.replace(/\./g, ",").replace(/,([^,]*)$/, ".$1");
    } else if (singleComma) {
      return numberPart.replace(",", ".");
    }

    return numberPart;
  }

  /**
   * Generate standardized error message for European format detection.
   *
   * @param {string} originalInput - The original user input
   * @param {string} ukSuggestion - The UK format suggestion
   * @returns {string} Formatted error message
   * @private
   */
  _generateEuropeanFormatError(originalInput, ukSuggestion) {
    return `Unsupported number format: '${originalInput}'. ` +
           `Please use: '${ukSuggestion}'. ` +
           "Kigali Sim requires comma for thousands separator and period for decimal point.";
  }
}

export {NumberParseUtil, NumberParseResult};
